### DISCUSSION TITLE
New papis neovim plugin

### DISCUSSION THREAD

**Discussion by jghauser** [papis/papis#388] [Ideas]
I'm thinking about writing a new plugin for neovim. I've got some experience writing lua plugins using the new apis and I thought I'd give this a shot in the near future.

At the moment I have one main question: How do I get queries to be reasonably fast? My desired core functionality would be akin to [telescope-bibtex](https://github.com/nvim-telescope/telescope-bibtex.nvim/), where you open a [`telescope`](https://github.com/nvim-telescope/telescope.nvim) floating window with a command/keybind and then input your query. Using `papis list` like in the existing `papis.vim` plugin is just faaar too slow for this use case. On my system, it takes a good 7ish seconds to open. `telescope-bibtex` which parses my bibtex file (which includes all bibliography entries) does the same thing almost instantaneously. This leads me to think I should implement some persistent database specifically for my plugin and update that database whenever something new is added to papis. It still doesn't really feel like an optimal solution...

**Comment by jghauser** [comment]
And of course, if anyone has ideas regarding features, feel free to add them here!

So far I plan on the following:
- (the above described) telescope search (from which it would be possible to insert the reference (citation key or full entry), open associated files, notes, and edit the yaml)
- displaying info about the entry when cursor is on the citation key
- opening associated files and notes when cursor is on the citation key
- a completion source for citation keys (lower priority as I don't really add them like this)

**Comment by alejandrogallo** [comment]
Hi! ok perfect,

there are some different ways of doing that I think, but I don't know which is the best.

# Exporting the bibtex file

The way I do this now in emacs with `org-ref` or `org-cite` is by simply exporting
everytime I do a commit in my papis library the bibtex file, ie, I have the following

```sh
#!/bin/sh
#
# An example hook script to verify what is about to be committed.
# Called by "git commit" with no arguments.  The hook should
# exit with non-zero status after issuing an appropriate message if
# it wants to stop the commit.
#
# To enable this hook, rename this file to "pre-commit".
rm ~/Documents/papers/bibfile.bib
papis \
  export \
  --all \
  --format bibtex -o ~/Documents/papers/bibfile.bib
(cd ~/Documents/papers/; git add bibfile.bib)
```

in the `.git/hooks/pre-commit` file.
In emacs I simply have the relevant package linked to this bib file.

An additional field is exported to the bibtex entry
with the full path to the papis folder of the entry, maybe
you can use this path to issue papis commands like
```
papis open --doc-folder {path}
```
and so on.

# Server

Another way would be to use the **incomplete** json API
in `papis serve`, this hoewever would need still some work to be completed,
but it would go around the problem of being slow.

# The way to go

I think for now probably the exporting to bibtex is the most realistic alternative and the one
where you ahve to work the least because you can reuse codes that show and
read the bibtex files. But it would be good to have some discussion

**Reply by jghauser** [reply]
Thanks a lot for your quick reply @alejandrogallo ! I wasn't very happy with the bibtex parsing implemented by the existing plugin and rather than writing a more complete parser myself, I opted to use the existing lyaml module and parse the yaml files with it. This has been going quite smoothly. I'll be working on and off on this and will post more questions here as I encounter them. The whole thing will probably take a while...

**Comment by jghauser** [comment]
hey @alejandrogallo! I've been slowly making my way to a first release of the neovim plugin (it's almost done!). There's one more feature I'd like to add at the moment, namely the ability of formatting a new note in a certain way when it is first created. I know that this feature has recently been merged into papis, but I'd still like to replicate it in the plugin, mostly because formatting with the help of lua and neovim is a lot more powerful/easy than even jinja. What I imagine doing is having the user be able to supply a custom lua function that will be applied to all newly created notes.

My problem is that I haven't found a way of using papis to create a note *without* also then opening the note. Is this possible? If not, do you have any hints of how accomplish it with a short python script? I could probably find out myself, but I find the papis codebase a little intimidating at times :sweat_smile:. Thanks a lot!

**Reply by alexfikl** [reply]
What were you trying to use to handle notes from `papis`? Do you need to go through `papis` for this?

In https://github.com/papis/papis/discussions/388#discussioncomment-3004491, you said you're parsing the `info.yaml` files and using that to get the document info. For the notes handling, you would just need to
* add an `notes: my-notes-file.rst` entry to the `info.yaml` file
* create the `my-notes-file.rst` in the document directory with whatever pre- or post-processing you need.

I don't think (should probably check) that `papis` does any other fancy handling of notes files, so that should be a good first step.

**Reply by jghauser** [reply]
My thinking is basically that I'm trying to do all things that change the papis files through the papis command, so that the plugin has less opportunities to mess things up. But that's not a hard rule if doing it otherwise is much simpler.

I would also need to somehow update the papis database, right? If I just add the "tag" key to the yaml, papis wouldn't know about it, right?

EDIT: I do read the yaml files with a lua library i found. I could also write to them using the same library, so it wouldn't be too hard to do that way.

**Reply by alexfikl** [reply]
> I would also need to somehow update the papis database, right? If I just add the "tag" key to the yaml, papis wouldn't know about it, right?

Hm, that's a good point. The `edit` command also does that
https://github.com/papis/papis/blob/35564265812fa6617222ad78bac4a3f75a765750/papis/commands/edit.py#L72-L100
i.e. if the entry doesn't exist, it creates a notes file, updates the db and opens it.

**Reply by jghauser** [reply]
Thanks! Yeah, I was trying to do it with the `edit` command, but the file opening isn't what I want. And my guess is that in order to update the db I would have to write a quick python script anyway, so I might as well do something that pretty much clones the `edit` behaviour just without opening the file (of course it would be even nicer if papis could do it natively).

**Reply by alexfikl** [reply]
One thing you can do is (or the equivalent in code)
```
papis update --set notes notes-file.rst [query]
```
to set and update the db with papis and then create the notes file from lua. Would that work?

**Reply by jghauser** [reply]
Yes, I think that should work! Thanks! I'll give it a shot and report back :)

**Reply by jghauser** [reply]
Do you know if it's possible to have papis use the defined `notes-name` when using `--set`? That would be nice as it would allow consistency between names (and consequenctly file formats) of the files generated by my plugin and papis itself. I can of course query papis for the `notes-name` but that would end up with garbage for those (like me) who have the filename dynamically generated by papis.

**Reply by alexfikl** [reply]
Hm, as far as I know that's not possible at the moment. 

The `--set` thing just updates the value in the document, but doesn't perform any processing on it. That dynamic part seems to only be handled when using `papis edit -n`. You'll probably need to add some similar logic to `papis update`.

**Reply by jghauser** [reply]
Thanks, that's what I thought. I'll see what seems like the best way to handle this

**Reply by jghauser** [reply]
Actually, I think passing a string that defines the dynamic formatting to `--set` does work, except the string isn't cleaned. I'm gonna try figure out what I need to change in papis to get that to work and then it should be as easy as querying `notes-name` and passing that to `--set`. :partying_face:

**Reply by alejandrogallo** [reply]
Sorry for the delay julian, isn't `papis config notes-name` what you're looking for?

**Reply by jghauser** [reply]
No worries @alejandrogallo, I think this has all been resolved. What I needed is a way of creating new notes and for people (like me) who have them dynamically generated by patterns (I use the jinja2 stuff for it), just getting the notes-name won't do. But after a small PR that ensures that notes created with `--set` are properly cleaned, it now all works :).

**Comment by jghauser** [comment]
I have published a wip version of [papis.nvim](https://github.com/jghauser/papis.nvim) for those interested in trying it out. very happy about any suggestions, feedback, bug reports, and so on. :)
